File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hashmap = {}

        for word in strs:
            new_word = ''.join(sorted(word))
            if new_word not in hashmap.keys():
                hashmap[new_word] = [word]
            else:
                hashmap[new_word] += [word]
    
        return list(hashmap.values())

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap<>();

        for (String word: strs) {
            String sortedWord = sortString(word);
            map.putIfAbsent(sortedWord, new ArrayList<>());
            map.get(sortedWord).add(word);
        }

        return convertToList(map);
    }


    private String sortString(String word) {
        char[] characters = word.toCharArray();
        Arrays.sort(characters);
        return new String(characters);
    }


    private List<List<String>> convertToList(HashMap<String, List<String>> map) {
        List<List<String>> result = new ArrayList<>();

        for (List<String> values: map.values()) {
            result.add(values);
        }

        return result;
    }
}